environment/src/main/java/com/sequenceiq/environment/environment/service/DistroxService.java

```diff
@@ -0,0 +1,77 @@
+import javax.ws.rs.WebApplicationException;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import com.dyngr.Polling;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.StackViewV4Response;
+import com.sequenceiq.distrox.api.v1.distrox.endpoint.DistroXV1Endpoint;
+
+@Service
+public class DistroxService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DistroxService.class);
+
+    @Value("${env.stop.polling.attempt:90}")
+    private Integer attempt;
+
+    @Value("${env.stop.polling.sleep.time:5}")
+    private Integer sleeptime;
+
+    private final DistroXV1Endpoint distroXV1Endpoint;
+
+    private final PollerCollection pollerCollection;
+
+    public DistroxService(DistroXV1Endpoint distroXV1Endpoint, PollerCollection pollerCollection) {
+        this.distroXV1Endpoint = distroXV1Endpoint;
+        this.pollerCollection = pollerCollection;
+    }
+
+    public Collection<StackViewV4Response> getAttachedDistroXClusters(String environmentName, String environmentCrn) {
+        LOGGER.debug("Get DistroX clusters of the environment: '{}'", environmentName);
+        try {
+            return distroXV1Endpoint.list(environmentName, environmentCrn).getResponses();
+        } catch (WebApplicationException | ProcessingException e) {
+            String message = String.format("Failed to get DistroX clusters from service due to: '%s' ", e.getMessage());
+            LOGGER.error(message, e);
+            throw e;
+        }
+    }
+
+    public void stopAttachedDistrox(Long envId, String envName) {
+        Collection<StackViewV4Response> attachedDistroXClusters = getAttachedDistroXClusters(envName, null);
+        ArrayList<String> pollingCrn = mapStackToCrn(attachedDistroXClusters);
+        distroXV1Endpoint.putStopByCrns(pollingCrn);
+        Polling.stopAfterAttempt(attempt)
+                .stopIfException(true)
+                .waitPeriodly(sleeptime, TimeUnit.SECONDS)
+                .run(pollerCollection.stopDistroXPoller(pollingCrn, envId));
+    }
+
+    public void startAttachedDistrox(Long envId, String envName) {
+        Collection<StackViewV4Response> attachedDistroXClusters = getAttachedDistroXClusters(envName, null);
+        ArrayList<String> pollingCrn = mapStackToCrn(attachedDistroXClusters);
+        distroXV1Endpoint.putStartByCrns(pollingCrn);
+        Polling.stopAfterAttempt(attempt)
+                .stopIfException(true)
+                .waitPeriodly(sleeptime, TimeUnit.SECONDS)
+                .run(pollerCollection.startDistroXPoller(pollingCrn, envId));
+
+    }
+
+    @NotNull
+    private ArrayList<String> mapStackToCrn(Collection<StackViewV4Response> attachedDistroXClusters) {
+        return attachedDistroXClusters.stream().map(StackViewV4Response::getCrn).collect(Collectors.toCollection(ArrayList::new));
```

Comment:
Collectors.toList();

Reply:
I need the exact ArrayList implementation, but the `Collectors.toList();` has another return type

Reply:
You should use set instead of this list implementation. You need uniqueness, but don't need ordering. See my other comment.

Reply:
The main problem is modifiable or unmodifiable implementations. I remove values from this list. The list or set declaration do not solve my problem. The most of factory methods are providing an unmodifiable list (like Collectors.toList(), Arrays.asList(), List.of()) and it will break my code in runtime. This is true for the Set as well. So I would use `ArraytList`.

Reply:
You are propagating ArrayList dependency unnecessarily. You should use the Set interface, and for modifications, wrap them into modifiable Sets at the last step.

Reply:
I don't want to use an unnecessary wrap, it is fully internal in this class